backend/app/Actions/Visits/CreateVisitAction.php

```diff
@@ -0,0 +1,181 @@
+        );
+
+        $geoPosition = $this->getOrCreateGeoPosition($request->ip());
+
+        $session = $this->getOrCreateSession(
+            $visitor->id,
+            $page->id,
+            $request->language(),
+            $system->id
+        );
+
+        $visit = new Visit();
+
+        $visit->visit_time = Carbon::now();
+        $visit->ip_address = $request->ip();
+        $visit->session_id = $session->id;
+        $visit->page_id = $page->id;
+        $visit->visitor_id = $visitor->id;
+        $visit->geo_position_id = $geoPosition->id;
+
+        $this->visitRepository->save($visit);
+    }
+
+    private function getOrCreatePage(int $websiteId, string $pageTitle, string $pageUrl): Page
+    {
+        $page = $this->pageRepository->getByParameters($websiteId, $pageTitle, $pageUrl);
+
+        if ($page !== null) {
+            return $page;
+        }
+
+        $page = new Page();
+
+        $page->name = $pageTitle;
+        $page->url = $pageUrl;
+        $page->previews = 0;
+        $page->website_id = $websiteId;
+
+        return $this->pageRepository->save($page);
+    }
+
+    private function getOrCreateSystem(
+        string $operatingSystem,
+        string $device,
+        string $browser,
+        int $resolutionHeight,
+        int $resolutionWidth
+    ): System {
+        $system = $this->systemRepository->getByParameters(
+            $operatingSystem,
+            $device,
+            $browser,
+            $resolutionHeight,
+            $resolutionWidth
+        );
+
+        if ($system !== null) {
+            return $system;
+        }
+
+        $system = new System();
+
+        $system->os = $operatingSystem;
+        $system->device = $device;
+        $system->browser = $browser;
+        $system->resolution_height = $resolutionHeight;
+        $system->resolution_width = $resolutionWidth;
+
+        return $this->systemRepository->save($system);
+    }
+
+    private function getOrCreateGeoPosition(string $ip): GeoPosition
+    {
+        $location = geoip($ip);
+        $geoPosition = $this->geoPositionRepository->getByParameters($location->country, $location->city);
+
+        if ($geoPosition !== null) {
+            return $geoPosition;
+        }
+
+        $geoPosition = new GeoPosition();
+
+        $geoPosition->country = $location->country;
+        $geoPosition->city = $location->city;
+
+        return $this->geoPositionRepository->save($geoPosition);
+    }
+
+    private function getOrCreateSession(
+        int $visitorId,
+        int $pageId,
+        string $language,
+        int $systemId
+    ): Session {
+        $session = $this->sessionRepository->lastActiveByVisitorId($visitorId);
+
+        if ($session !== null) {
+            $this->sessionRepository->updateEndSession($session);
+            return $session;
+        }
+
+        $session = new Session();
+
+        $session->start_session = Carbon::now();
+        $session->visitor_id = $visitorId;
```

Comment:
end_session also should be set, as it points on last activity of session. Queries won't work for first session with one visit